ztest: drivers: can: add imx93_evk to drivers/can/api test

CkovMk opened the pull request:
Add imx93_evk to drivers/can/api test

Link to issue: #73757

alexanderwachter commented on tests/drivers/can/api/board/mimx93_evk_a55.overlay:
```diff
@@ -0,0 +1,28 @@
+};
+
+&flexcan2 {
+	bus-speed = <1000000>;
+    bus-speed-data = <5000000>;
```
```suggestion
```

These properties are not needed anymore. The tree wide default is set via Kconfig.

CkovMk replied:
removed

alexanderwachter commented on tests/drivers/can/api/board/mimx93_evk_a55.overlay:
```diff
@@ -0,0 +1,28 @@
+    };
+};
+
+&lpi2c2 {
+    status = "okay";
+};
+
+&mfd0 {
+    status = "okay";
+};
+
+&gpiohog_exp_sel {
+    status = "okay";
+    output-low;
+};
```
How do this change relate to CAN?

CkovMk replied:
this gpio controls a series of muxes on the board. pull it low yo select can signal.

henrikbrixandersen reviewed (changes requested):
I fail to understand why an overlay is needed here - isn't the board configured to use CAN already?

Also, how does this PR fix #73757?

pdgendt commented on tests/drivers/can/api/boards/imx93_evk_mimx9352_a55.overlay:
Use tabs for indentation instead of spaces

CkovMk replied:
sure.

CkovMk commented:
> I fail to understand why an overlay is needed here - isn't the board configured to use CAN already?
> 
> Also, how does this PR fix #73757?

It did not fix, just to show overlay needed to run this test.

henrikbrixandersen commented:
> I was told to disable related nodes in board dts to keep it clean, so that end developer could only enable nodes they need. Also, a few KConfig options are not enabled in board defconfig.

Told by whom?

CkovMk commented:
> > I was told to disable related nodes in board dts to keep it clean, so that end developer could only enable nodes they need. Also, a few KConfig options are not enabled in board defconfig.
> 
> Told by whom?

@JiafeiPan maybe we should discuss here, which node(s) should be enabled by default?

JiafeiPan commented:
> > > I was told to disable related nodes in board dts to keep it clean, so that end developer could only enable nodes they need. Also, a few KConfig options are not enabled in board defconfig.
> > 
> > 
> > Told by whom?
> 
> @JiafeiPan maybe we should discuss here, which node(s) should be enabled by default?

As we can run multiple OS on each CPU Core of MPU platform, for example we need to run Zephyr together with Linux, so by default I think we should keep a minimal enabled devices in dts for Zephyr to avoid conflict with Linux ( I know we should to disable the device enabled in Zephyr, but I don;t want to disable CAN in Zephyr dts when I just want to run a hello world on Zephyr). so except basic peripherals (such as CPU Core, GIC, CCM, pinmux, UART), most of the other peripherals should be disabled by default in Zephyr dts, we can enabled them as needed by overlay in application/usecases.

JiafeiPan commented:
You should disable dts node which enabled in https://github.com/zephyrproject-rtos/zephyr/pull/70201 firstly, then enable them in overlay.

JiafeiPan reviewed (changes requested):
You should disable dts node which enabled in https://github.com/zephyrproject-rtos/zephyr/pull/70201 firstly, then enable them in overlay.

CkovMk commented:
@JiafeiPan this should do the job.

JiafeiPan commented on samples/drivers/can/counter/boards/imx93_evk_mimx9352_a55.conf:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_I2C=y
+CONFIG_MFD=y
```
why CONFIG_GPIO is not enabled at this case but enabled with all the other case?

CkovMk replied:
it was enabled in prj.conf in this particular project.

JiafeiPan commented on samples/drivers/can/babbling/boards/imx93_evk_mimx9352_a55.conf:
```diff
@@ -0,0 +1,6 @@
+
+CONFIG_I2C=y
+CONFIG_MFD=y
+CONFIG_GPIO_HOGS=y
```
we don't need to enable all these items, they are enabled by dts by default.

CkovMk replied:
so conf is not needed

CkovMk committed:
ztest: can: add imx93_evk to can test and samples

Add imx93_evk to can test and samples

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>

henrikbrixandersen commented:
Come one guys, this doesn't scale.

Enabling devicetree nodes in the board configuration is "free" until an application enables the corresponding Kconfig options. All devicetree nodes, that are supported by a board, should be enabled in the board devicetree files.

JiafeiPan commented:
> Come one guys, this doesn't scale.
> 
> Enabling devicetree nodes in the board configuration is "free" until an application enables the corresponding Kconfig options. All devicetree nodes, that are supported by a board, should be enabled in the board devicetree files.

Clear now, my misunderstand, thanks.

CkovMk commented:
@henrikbrixandersen How do you like this solution?

Having talked with @hakehuang and @JiafeiPan , I believe the more proper way to do this is still using overlay. To make it cleaner, I created overlay for the board and include them in tests and samples. 

Here's the reason to do so:

The EXP_SEL signal from ADP5585 (the &mfd0 node) gpio expander controls a set of digital muxes that switches the pin functionalities. Different function may require different state of this signal. As you stated:

> All devicetree nodes, that are supported by a board, should be enabled in the board devicetree files.

It is not the simple case here: some devicetree nodes that can be supported by this board can NOT be enabled simultaneously. Part of them can only be enabled when EXP_SEL is HIGH, and part of them can only be enabled when EXP_SEL is LOW.

For example, LPI2C8 on 40pin Rpi header is conflict with FlexCAN2.

To avoid future conflict, I strongly recommend we should use overlay here.

CkovMk commented:
@henrikbrixandersen How do you like this solution?

Having talked with @hakehuang and @JiafeiPan , I believe the more proper way to do this is still using overlay. To make it cleaner, I created overlay for the board and include them in tests and samples. 

Here's the reason to do so:

The EXP_SEL signal from ADP5585 (the &mfd0 node) gpio expander controls a set of digital muxes that switches the pin functionalities. Different function may require different state of this signal. As you stated:

> All devicetree nodes, that are supported by a board, should be enabled in the board devicetree files.

It is not the simple case here: some devicetree nodes that can be supported by this board can NOT be enabled simultaneously. Part of them can only be enabled when EXP_SEL is HIGH, and part of them can only be enabled when EXP_SEL is LOW.

For example, LPI2C8 on 40pin Rpi header is conflict with FlexCAN2.

To avoid future conflict, I strongly recommend we should use overlay here.

CkovMk commented:
tests/drivers/can/api runs on i.MX93 EVK:

```
*** Booting Zephyr OS build v3.6.0-5870-g8cb3625982b6 ***
Running TESTSUITE can_classic
===================================================================
START - test_add_filter
 PASS - test_add_filter in 0.001 seconds
===================================================================
START - test_add_invalid_ext_filter
E: invalid filter with extended (29-bit) CAN ID 0x1fffffff, CAN ID mask 0x20000000
E: invalid filter with extended (29-bit) CAN ID 0x20000000, CAN ID mask 0x1fffffff
 PASS - test_add_invalid_ext_filter in 0.015 seconds
===================================================================
START - test_add_invalid_null_filter
 PASS - test_add_invalid_null_filter in 0.001 seconds
===================================================================
START - test_add_invalid_std_filter
E: invalid filter with standard (11-bit) CAN ID 0x7ff, CAN ID mask 0x800
E: invalid filter with standard (11-bit) CAN ID 0x800, CAN ID mask 0x7ff
 PASS - test_add_invalid_std_filter in 0.013 seconds
===================================================================
START - test_bitrate_limits
 PASS - test_bitrate_limits in 0.001 seconds
===================================================================
START - test_classic_get_capabilities
 PASS - test_classic_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_added_while_stopped
 PASS - test_filters_added_while_stopped in 0.002 seconds
===================================================================
START - test_filters_preserved_through_bitrate_change
 PASS - test_filters_preserved_through_bitrate_change in 0.003 seconds
===================================================================
START - test_filters_preserved_through_mode_change
 PASS - test_filters_preserved_through_mode_change in 0.003 seconds
===================================================================
START - test_get_core_clock
 PASS - test_get_core_clock in 0.001 seconds
===================================================================
START - test_get_state
 PASS - test_get_state in 0.001 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_max_ext_filters
 PASS - test_max_ext_filters in 0.001 seconds
===================================================================
START - test_max_std_filters
 PASS - test_max_std_filters in 0.001 seconds
===================================================================
START - test_receive_timeout
 PASS - test_receive_timeout in 0.101 seconds
===================================================================
START - test_recover
 PASS - test_recover in 0.001 seconds
===================================================================
START - test_recover_while_stopped
 PASS - test_recover_while_stopped in 0.001 seconds
===================================================================
START - test_reject_ext_id_rtr
 PASS - test_reject_ext_id_rtr in 0.101 seconds
===================================================================
START - test_reject_std_id_rtr
 PASS - test_reject_std_id_rtr in 0.101 seconds
===================================================================
START - test_send_and_forget
 PASS - test_send_and_forget in 0.001 seconds
===================================================================
START - test_send_callback
 PASS - test_send_callback in 0.001 seconds
===================================================================
START - test_send_ext_id_out_of_range
E: invalid frame with extended (29-bit) CAN ID 0x20000000
 PASS - test_send_ext_id_out_of_range in 0.006 seconds
===================================================================
START - test_send_fd_format
E: unsupported CAN frame flags 0x04
 PASS - test_send_fd_format in 0.004 seconds
===================================================================
START - test_send_invalid_dlc
E: DLC of 9 exceeds maximum (8)
 PASS - test_send_invalid_dlc in 0.003 seconds
===================================================================
START - test_send_null_frame
 PASS - test_send_null_frame in 0.001 seconds
===================================================================
START - test_send_receive_ext_id
 PASS - test_send_receive_ext_id in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_masked
 PASS - test_send_receive_ext_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_ext_id_rtr
 SKIP - test_send_receive_ext_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_msgq
 PASS - test_send_receive_msgq in 0.010 seconds
===================================================================
START - test_send_receive_std_id
 PASS - test_send_receive_std_id in 0.004 seconds
===================================================================
START - test_send_receive_std_id_masked
 PASS - test_send_receive_std_id_masked in 0.004 seconds
===================================================================
START - test_send_receive_std_id_no_data
 PASS - test_send_receive_std_id_no_data in 0.001 seconds
===================================================================
START - test_send_receive_std_id_rtr
 SKIP - test_send_receive_std_id_rtr in 0.001 seconds
===================================================================
START - test_send_receive_wrong_id
 PASS - test_send_receive_wrong_id in 0.102 seconds
===================================================================
START - test_send_std_id_out_of_range
E: invalid frame with standard (11-bit) CAN ID 0x800
 PASS - test_send_std_id_out_of_range in 0.005 seconds
===================================================================
START - test_send_while_stopped
 PASS - test_send_while_stopped in 0.001 seconds
===================================================================
START - test_set_bitrate
 PASS - test_set_bitrate in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_bitrate_while_started
 PASS - test_set_bitrate_while_started in 0.001 seconds
===================================================================
START - test_set_mode_while_started
 PASS - test_set_mode_while_started in 0.001 seconds
===================================================================
START - test_set_state_change_callback
 PASS - test_set_state_change_callback in 0.001 seconds
===================================================================
START - test_set_timing_max
 PASS - test_set_timing_max in 0.009 seconds
===================================================================
START - test_set_timing_min
 PASS - test_set_timing_min in 0.002 seconds
===================================================================
START - test_set_timing_while_started
 PASS - test_set_timing_while_started in 0.001 seconds
===================================================================
START - test_start_while_started
 PASS - test_start_while_started in 0.001 seconds
===================================================================
START - test_stop_while_stopped
 PASS - test_stop_while_stopped in 0.001 seconds
===================================================================
TESTSUITE can_classic succeeded
Running TESTSUITE can_stats
===================================================================
START - test_can_stats_accessors
 PASS - test_can_stats_accessors in 0.001 seconds
===================================================================
TESTSUITE can_stats succeeded
Running TESTSUITE can_transceiver
===================================================================
START - test_get_transceiver
 PASS - test_get_transceiver in 0.001 seconds
===================================================================
TESTSUITE can_transceiver succeeded
Running TESTSUITE can_utilities
===================================================================
START - test_can_bytes_to_dlc
 PASS - test_can_bytes_to_dlc in 0.001 seconds
===================================================================
START - test_can_dlc_to_bytes
 PASS - test_can_dlc_to_bytes in 0.001 seconds
===================================================================
START - test_can_frame_matches_filter
 PASS - test_can_frame_matches_filter in 0.001 seconds
===================================================================
TESTSUITE can_utilities succeeded
Running TESTSUITE canfd
===================================================================
START - test_canfd_get_capabilities
 PASS - test_canfd_get_capabilities in 0.001 seconds
===================================================================
START - test_filters_preserved_through_classic_to_fd_mode_change

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:101: send_test_frame: (err not equal to 0)
failed to send frame (err -114)
 FAIL - test_filters_preserved_through_classic_to_fd_mode_change in 0.015 seconds
===================================================================
START - test_filters_preserved_through_fd_to_classic_mode_change

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:101: send_test_frame: (err not equal to 0)
failed to send frame (err -114)
 FAIL - test_filters_preserved_through_fd_to_classic_mode_change in 0.015 seconds
===================================================================
START - test_invalid_sample_point
 PASS - test_invalid_sample_point in 0.001 seconds
===================================================================
START - test_send_receive_classic

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:161: add_rx_filter: (filter_id equal to -ENOSPC)
no filters available
 FAIL - test_send_receive_classic in 0.014 seconds
===================================================================
START - test_send_receive_fd

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:137: add_rx_msgq: (filter_id equal to -ENOSPC)
no filters available
 FAIL - test_send_receive_fd in 0.013 seconds
===================================================================
START - test_send_receive_mixed

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/can/api/src/canfd.c:137: add_rx_msgq: (filter_id equal to -ENOSPC)
no filters available
 FAIL - test_send_receive_mixed in 0.013 seconds
===================================================================
START - test_set_bitrate_data_while_started
 PASS - test_set_bitrate_data_while_started in 0.001 seconds
===================================================================
START - test_set_bitrate_too_high
 PASS - test_set_bitrate_too_high in 0.001 seconds
===================================================================
START - test_set_timing_data_max
 PASS - test_set_timing_data_max in 0.001 seconds
===================================================================
START - test_set_timing_data_min
 PASS - test_set_timing_data_min in 0.001 seconds
===================================================================
START - test_set_timing_data_while_started
 PASS - test_set_timing_data_while_started in 0.001 seconds
===================================================================
TESTSUITE canfd failed.

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [can_classic]: pass = 44, fail = 0, skip = 2, total = 46 duration = 0.522 seconds
 - PASS - [can_classic.test_add_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_ext_filter] duration = 0.015 seconds
 - PASS - [can_classic.test_add_invalid_null_filter] duration = 0.001 seconds
 - PASS - [can_classic.test_add_invalid_std_filter] duration = 0.013 seconds
 - PASS - [can_classic.test_bitrate_limits] duration = 0.001 seconds
 - PASS - [can_classic.test_classic_get_capabilities] duration = 0.001 seconds
 - PASS - [can_classic.test_filters_added_while_stopped] duration = 0.002 seconds
 - PASS - [can_classic.test_filters_preserved_through_bitrate_change] duration = 0.003 seconds
 - PASS - [can_classic.test_filters_preserved_through_mode_change] duration = 0.003 seconds
 - PASS - [can_classic.test_get_core_clock] duration = 0.001 seconds
 - PASS - [can_classic.test_get_state] duration = 0.001 seconds
 - PASS - [can_classic.test_invalid_sample_point] duration = 0.001 seconds
 - PASS - [can_classic.test_max_ext_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_max_std_filters] duration = 0.001 seconds
 - PASS - [can_classic.test_receive_timeout] duration = 0.101 seconds
 - PASS - [can_classic.test_recover] duration = 0.001 seconds
 - PASS - [can_classic.test_recover_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_reject_ext_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_reject_std_id_rtr] duration = 0.101 seconds
 - PASS - [can_classic.test_send_and_forget] duration = 0.001 seconds
 - PASS - [can_classic.test_send_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_send_ext_id_out_of_range] duration = 0.006 seconds
 - PASS - [can_classic.test_send_fd_format] duration = 0.004 seconds
 - PASS - [can_classic.test_send_invalid_dlc] duration = 0.003 seconds
 - PASS - [can_classic.test_send_null_frame] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_ext_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_ext_id_masked] duration = 0.004 seconds
 - SKIP - [can_classic.test_send_receive_ext_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_msgq] duration = 0.010 seconds
 - PASS - [can_classic.test_send_receive_std_id] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_masked] duration = 0.004 seconds
 - PASS - [can_classic.test_send_receive_std_id_no_data] duration = 0.001 seconds
 - SKIP - [can_classic.test_send_receive_std_id_rtr] duration = 0.001 seconds
 - PASS - [can_classic.test_send_receive_wrong_id] duration = 0.102 seconds
 - PASS - [can_classic.test_send_std_id_out_of_range] duration = 0.005 seconds
 - PASS - [can_classic.test_send_while_stopped] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [can_classic.test_set_bitrate_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_mode_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_set_state_change_callback] duration = 0.001 seconds
 - PASS - [can_classic.test_set_timing_max] duration = 0.009 seconds
 - PASS - [can_classic.test_set_timing_min] duration = 0.002 seconds
 - PASS - [can_classic.test_set_timing_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_start_while_started] duration = 0.001 seconds
 - PASS - [can_classic.test_stop_while_stopped] duration = 0.001 seconds

SUITE PASS - 100.00% [can_stats]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_stats.test_can_stats_accessors] duration = 0.001 seconds

SUITE PASS - 100.00% [can_transceiver]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
 - PASS - [can_transceiver.test_get_transceiver] duration = 0.001 seconds

SUITE PASS - 100.00% [can_utilities]: pass = 3, fail = 0, skip = 0, total = 3 duration = 0.003 seconds
 - PASS - [can_utilities.test_can_bytes_to_dlc] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_dlc_to_bytes] duration = 0.001 seconds
 - PASS - [can_utilities.test_can_frame_matches_filter] duration = 0.001 seconds

SUITE FAIL -  58.33% [canfd]: pass = 7, fail = 5, skip = 0, total = 12 duration = 0.077 seconds
 - PASS - [canfd.test_canfd_get_capabilities] duration = 0.001 seconds
 - FAIL - [canfd.test_filters_preserved_through_classic_to_fd_mode_change] duration = 0.015 seconds
 - FAIL - [canfd.test_filters_preserved_through_fd_to_classic_mode_change] duration = 0.015 seconds
 - PASS - [canfd.test_invalid_sample_point] duration = 0.001 seconds
 - FAIL - [canfd.test_send_receive_classic] duration = 0.014 seconds
 - FAIL - [canfd.test_send_receive_fd] duration = 0.013 seconds
 - FAIL - [canfd.test_send_receive_mixed] duration = 0.013 seconds
 - PASS - [canfd.test_set_bitrate_data_while_started] duration = 0.001 seconds
 - PASS - [canfd.test_set_bitrate_too_high] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_max] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_min] duration = 0.001 seconds
 - PASS - [canfd.test_set_timing_data_while_started] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
RunID: ec19171ddf8a9bf8859efad71f70a9f4
PROJECT EXECUTION FAILED
```

CAN-FD test fail is marked as an known issue.

hakehuang commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -100,17 +99,18 @@
 
 &lpi2c2 {
-	status = "okay";
+	status = "disabled";
```
why still disable this?

CkovMk replied:
we can have this enabled by default. updated.

hakehuang commented on boards/nxp/imx93_evk/imx93_evk_mimx9352_a55.dts:
```diff
@@ -100,17 +99,18 @@
 	pinctrl-names = "default";
 
 	mfd0:adp5585@34 {
+		status = "disabled";
```
if we move the modifcations ito the imx93_evk_flexcan2.overlay, this should keep as is.

CkovMk committed:
board: imx93_evk: disable flexcan related nodes by default

Disable flexcan related nodes by default to prevent build issue.
Enable common nodes.

Signed-off-by: Chekhov Ma <chekhov.ma@nxp.com>
